applications/StructuralMechanicsApplication/custom_constitutive/linear_j2_plasticity_3d.cpp

```diff
@@ -174,95 +180,123 @@ void LinearJ2Plasticity3D::CalculateMaterialResponseKirchhoff(ConstitutiveLaw::P
-            stress_trial_dev(1) - 2. * mu * dgamma * yield_function_normal_vector(1);
-        stress_vector(2) =
-            volumetric_modulus * (strain_vector(0) + strain_vector(1) + strain_vector(2)) +
-            stress_trial_dev(2) - 2. * mu * dgamma * yield_function_normal_vector(2);
-        stress_vector(3) =
-            stress_trial_dev(3) - 2. * mu * dgamma * yield_function_normal_vector(3);
-        stress_vector(4) =
-            stress_trial_dev(4) - 2. * mu * dgamma * yield_function_normal_vector(4);
-        stress_vector(5) =
-            stress_trial_dev(5) - 2. * mu * dgamma * yield_function_normal_vector(5);
-
-        mPlasticStrain(0) = mPlasticStrainOld(0) + dgamma * yield_function_normal_vector(0);
-        mPlasticStrain(1) = mPlasticStrainOld(1) + dgamma * yield_function_normal_vector(1);
-        mPlasticStrain(2) = mPlasticStrainOld(2) + dgamma * yield_function_normal_vector(2);
-        mPlasticStrain(3) = mPlasticStrainOld(3) + dgamma * yield_function_normal_vector(3) * 2;
-        mPlasticStrain(4) = mPlasticStrainOld(4) + dgamma * yield_function_normal_vector(4) * 2;
-        mPlasticStrain(5) = mPlasticStrainOld(5) + dgamma * yield_function_normal_vector(5) * 2;
-        mAccumulatedPlasticStrain = mAccumulatedPlasticStrainOld + sqrt_two_thirds * dgamma;
+    // If we compute the tangent moduli or the stress
+    if( r_constitutive_law_options.Is( ConstitutiveLaw::COMPUTE_STRESS ) ||
+        r_constitutive_law_options.Is( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR )) {
+        Vector& r_stress_vector = rValues.GetStressVector();
+
+        if (rValues.GetProcessInfo().Has(INITIAL_STRAIN)) {
+            noalias(r_strain_vector) += rValues.GetProcessInfo()[INITIAL_STRAIN];
+        }
 
-        CalculateTangentTensor(dgamma, norm_dev_stress, yield_function_normal_vector,
-                               rMaterialProperties, tangent_tensor);
+        Matrix elastic_tensor;
+        Matrix& tangent_tensor = rValues.GetConstitutiveMatrix();
+        const double hardening_modulus = r_material_properties[ISOTROPIC_HARDENING_MODULUS];
+        const double delta_k = r_material_properties[INFINITY_HARDENING_MODULUS];
+        const double hardening_exponent = r_material_properties[HARDENING_EXPONENT];
+        const double E = r_material_properties[YOUNG_MODULUS];
+        const double poisson_ratio = r_material_properties[POISSON_RATIO];
+        const double mu = E / (2. + 2. * poisson_ratio);
+        const double volumetric_modulus = E / (3. * (1. - 2. * poisson_ratio));
+        const double sqrt_two_thirds = std::sqrt(2.0 / 3.0); // =0.8164965809277260
+        double trial_yield_function;
+
+        mPlasticStrain = mPlasticStrainOld;
+        mAccumulatedPlasticStrain = mAccumulatedPlasticStrainOld;
+
+        elastic_tensor.resize(6, 6, false);
+        CalculateElasticMatrix(elastic_tensor, r_material_properties);
+        Vector yield_tensionrial(6);
```

Comment:
Now that you are at it, could you correct my typo, from "yield_tensionrial" to "yield_tension"?

Reply:
I just saw, for a future PR